### PR TITLE
Add a jax backend based on dia sparse matrix, `JaxDia`.

### PR DESCRIPTION
**Pull request by Ericgig**
Add a dia sparse data layer for Jax.
`JaxDia`'s data are jax.array, (differentiable), but the structure is meta data.
This allow most functions to support jit.
`tidyup` does not jit, so I did not include automated tidyup.
The dia format is good for operator with few diagonals, passable for ket and horrible for bra.
I made the function so it can be supported as the operators in `mesolve`. 

Included specialisations:
- `isherm`, `iszero`, `isdiag`
- `add`, `sub`, `mul`, `matmul`, `multiply`, `kron`
- `neg`, `adjoint`, `transpose`, `conj`
- `zeros`, `identity`, `diag`, `one_element`
- `expect`, `expect_super`, `trace`
- `extract`, `tidyup`

Won't do:
- `indices`, `dimensions`, `reshape_jaxarray`, `column_stack_jaxarray`, `column_unstack_jaxarray`, `split_columns_jaxarray`, `trace_oper_ket`.
  Reshaping, permuting breaks the diagonals structure.
- `inv`, `expm`, `eigs`, `svd`, `solve`. Non-trivial, better done in dense format.
  
Maybe:
- `project`, `pow`, `inner`, `inner_op`, `norm...`,  `ptrace`

#19 is a branch off this one, and should be merged before this one.

### CODE REVIEW

File: src/qutip_jax/jaxdia.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import qutip.core.data as _data
+import numpy as np
+from qutip.core.data.base import Data
+import numbers
+
+
+config.update("jax_enable_x64", True)
+
+
+__all__ = ["JaxDia"]
```
Comment (Ericgig) [inline src/qutip_jax/jaxdia.py:15]:
```suggestion
config.update("jax_enable_x64", True)

__all__ = ["JaxDia"]
```

File: src/qutip_jax/binops.py [hunk 2]
```diff
@@ -108,12 +193,140 @@ def matmul_jaxarray(left, right, scale=1, out=None):
...
     return JaxArray._fast_constructor(left._jxa * right._jxa, shape=left.shape)
 
 
+@jit
+def multiply_jaxdia(left, right):
+    """
+    Perform the operation
+        left + scale*right
+    where `left` and `right` are matrices, and `scale` is an optional complex
+    scalar.
```
Comment (nwlambert) [inline src/qutip_jax/binops.py:309]: Seems like the wrong docstring here...

File: src/qutip_jax/binops.py [hunk 3]
```diff
@@ -122,6 +335,74 @@ def kron_jaxarray(left, right):
     return JaxArray(jnp.kron(left._jxa, right._jxa))
 
 
+@jit
+def _multiply_outer(left, right):
+    return vmap(vmap(jnp.multiply, (None, 0)), (0, None))(left, right).ravel()
+
+
+@jit
+def kron_jaxdia(left, right):
```
Comment (nwlambert) [inline src/qutip_jax/binops.py:344]: missing docstring?

File: src/qutip_jax/ode.py [hunk 4]
```diff
@@ -22,6 +22,14 @@ def _float2cplx(arr):
     return arr[0] + 1j * arr[1]
 
 
+@jax.jit
+def dstate(t, y, args):
+    state = _float2cplx(y)
+    H, = args
+    d_state = H.matmul_data(t, JaxArray(state))
+    return _cplx2float(d_state._jxa)
```
Comment (nwlambert) [inline src/qutip_jax/ode.py:30]: just a minor thing, does the new complex support in diffrax essentially do the same thing?  though given all the warnings using complex in diffrax directly throws out, maybe better to stick with this manual treatment at the moment.

**Review by nwlambert — APPROVED** [review]
Looks great, I stress tested mesolve() with this quite a bit in benchmarking, and didnt have any issues.  Left a couple of comments on wrong/missing docstrings.  

One random comment, more on qutip jax than jaxdia, would making the ode solver use diffraxs PIDController by default instead of ConstantStepSize cause any problems (assuming its still the default)  ? in practice the performance of the ConstantStepSize was terrible (for the cases I tried)

File: src/qutip_jax/ode.py (hunk 4 above)
Comment (Ericgig) [reply src/qutip_jax/ode.py:30]: No sure, I did not plan too much with it yet.

**Comment by Ericgig** [comment]
We can change to PIDController before making a release, but not in this PR.

**Review by nwlambert — APPROVED** [review]

**Pull request merged**